bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java

```diff
@@ -500,129 +498,68 @@ public void safeRun() {
-                                                ledgerId, BKException.codeLogger(newrc));
-                                        cb.closeComplete(rc, LedgerHandle.this, ctx);
+                    tearDownWriteHandleState();
+                    new MetadataUpdateLoop(
+                            clientCtx.getLedgerManager(), getId(),
+                            LedgerHandle.this::getLedgerMetadata,
+                            (metadata) -> {
+                                if (metadata.isClosed()) {
+                                    /* If the ledger has been closed with the same lastEntry
+                                     * and length that we planned to close with, we have nothing to do,
+                                     * so just return success */
+                                    if (lastEntry == metadata.getLastEntryId()
+                                        && finalLength == metadata.getLength()) {
+                                        return false;
                                     } else {
-                                        LedgerMetadata metadata = getLedgerMetadata();
-                                        metadata.setState(prevState);
-                                        if (prevState.equals(State.CLOSED)) {
-                                            metadata.close(prevLastEntryId);
-                                        }
-
-                                        metadata.setLength(prevLength);
-                                        if (!metadata.isNewerThan(newMeta)
-                                                && !metadata.isConflictWith(newMeta)) {
-                                            // use the new metadata's ensemble, in case re-replication already
-                                            // replaced some bookies in the ensemble.
-                                            metadata.setEnsembles(newMeta.getEnsembles());
-                                            metadata.setVersion(newMeta.version);
-                                            metadata.setLength(length);
-                                            metadata.close(getLastAddConfirmed());
-                                            writeLedgerConfig(new CloseCb());
-                                            return;
-                                        } else {
-                                            metadata.setLength(length);
-                                            metadata.close(getLastAddConfirmed());
-                                            LOG.warn("Conditional update ledger metadata for ledger {} failed.",
-                                                    ledgerId);
-                                            cb.closeComplete(rc, LedgerHandle.this, ctx);
-                                        }
+                                        throw new BKException.BKMetadataVersionException();
```

Comment:
Unless I'm mistaken, this can only happen (baring a bug) due to a recovery open.  Looks like the BKMetadataVersionException here gets translated to Code.MetadataVersionException in the whenComplete handler to be propagated to the caller.  It's a little obtuse for a condition the user might actually want to know about, but probably ok.  I, however, like to see a log line at ERROR if the racing metadata update ended up with a lastEntryId < LAC (which should be strictly impossible, should be a simple check here).